### PR TITLE
`[pylint]` Implement `len-test` (`PLC1802`)

### PR DESCRIPTION
**Pull request by Lokejoke**
## Summary

This PR implements [`use-implicit-booleaness-not-len` / `C1802`](https://pylint.pycqa.org/en/latest/user_guide/messages/convention/use-implicit-booleaness-not-len.html)
> For sequences, (strings, lists, tuples), use the fact that empty sequences are false. 

## Test Plan

Checked against pylint tests:
 * I would argue some of the tests (commented out) are not very intuitive and thus not helpful 

## Notes
* **Naming**: Rule name and its error message (adopted from pylint) do not follow conventions used in Ruff (`len-as-condition` might be solution) 

## References
[PEP 8 on empty sequences and implicit booleaness](https://peps.python.org/pep-0008/#programming-recommendations)

### CODE REVIEW

**Comment by dylwil3** [comment]
This looks very nice, thanks for the contribution! I'll try to review this more carefully in a bit but I was just glancing through the ecosystem results - do you know why the `bokeh` result has only one violation instead of several? It looks like `len(errors)` is flagged once but used several times in the surrounding snippet. (Apologies if I'm missing something - looking at this on my phone so may have gotten muddled)

**Review by dylwil3 — CHANGES_REQUESTED** [review]
Thanks so much for this contribution! Some minor nits/requests for a few more tests. I also didn't get a chance to review the implementation carefully, but I'll hold off for now in case the new tests inspire any changes.

File: crates/ruff_linter/resources/test/fixtures/pylint/len_as_condition.py
[the remark this answers is not in the document]
Comment (Lokejoke) [reply crates/ruff_linter/resources/test/fixtures/pylint/len_as_condition.py:1]: I've commented out the code i have mixed feelings about:
* object implicit boolean (first using '__bool__', then '__len__'), removing call might change the output
* than the function tests (maybe after stronger type inference)

I'll change it, so tests would show how they differ from pylint

[the remark this answers is not in the document]
Comment (Lokejoke) [reply crates/ruff_linter/resources/test/fixtures/pylint/len_as_condition.py:1]: I'm unsure, there are 3 options:
* check only those without rebindings (as is now)
* check recursively if all of possible rebindings bounds to a sequence-like object
* leave it for type inference

**Review by dylwil3 — CHANGES_REQUESTED** [review]
This looks really great, thank you! I think even with the possible misses around control flow, this makes sense to implement. You would have to be pretty pathological to create a custom class `C` that implements `__len__` and `__bool__` where `bool(x) != bool(len(x))` for `x: C`... so I think it's okay.

File: crates/ruff_linter/src/rules/pylint/rules/len_as_condition.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+fn is_sequence(expr: &Expr, semantic: &SemanticModel) -> bool {
+    // Check if the expression type is a direct sequence match (dict, list, set, tuple, generator, string)
+    if matches!(
+        ResolvedPythonType::from(expr),
+        ResolvedPythonType::Atom(
+            PythonType::Dict
+                | PythonType::List
+                | PythonType::Set
+                | PythonType::Tuple
+                | PythonType::Generator
```
[the remark this answers is not in the document]
Comment (Lokejoke) [reply crates/ruff_linter/src/rules/pylint/rules/len_as_condition.rs]: Yes, it throws an error. Generators should be excluded from the search.

**Review by dylwil3 — CHANGES_REQUESTED** [review]
Almost there! Looking good!

**Comment by MichaReiser** [comment]
Naming: Maybe `len-test` to align with https://docs.astral.sh/ruff/rules/not-in-test/#not-in-test-e713

**Review by dylwil3 — APPROVED** [review]
LGTM, thank you for the great contribution and your patience through all the reviews!

**Pull request merged**